activation/handler_v2.go

```diff
@@ -435,67 +390,211 @@ func (h *HandlerV2) validateMarriages(atx *wire.ActivationTxV2) ([]types.NodeID,
 		if err != nil {
 			return nil, nil, fmt.Errorf("fetching previous atx: %w", err)
 		}
-		if prevAtx.Published() >= atx.PublishEpoch {
-			err := fmt.Errorf("previous atx is too new (%d >= %d) (%s) ", prevAtx.Published(), atx.PublishEpoch, prev)
+		if prevAtx.PublishEpoch >= atx.PublishEpoch {
+			err := fmt.Errorf("previous atx is too new (%d >= %d) (%s) ", prevAtx.PublishEpoch, atx.PublishEpoch, prev)
 			return nil, nil, err
 		}
 		previousAtxs[i] = prevAtx
 	}
 
-	// validate all niposts
-	// TODO: support merged ATXs
-	// For a merged ATX we need to fetch the equivocation this smesher is part of.
-	equivocationSet := []types.NodeID{atx.SmesherID}
-	var totalEffectiveNumUnits uint32
-	var minLeaves uint64 = math.MaxUint64
-	var smesherCommitment *types.ATXID
-	for _, niposts := range atx.NiPosts {
-		// verify PoET memberships in a single go
-		var poetChallenges [][]byte
+	equivocationSet, err := h.equivocationSet(atx)
+	if err != nil {
+		return nil, nil, fmt.Errorf("calculating equivocation set: %w", err)
+	}
 
+	// validate previous ATXs
+	nipostSizes := make(nipostSizes, len(atx.NiPosts))
+	for i, niposts := range atx.NiPosts {
+		nipostSizes[i] = new(nipostSize)
 		for _, post := range niposts.Posts {
 			if post.MarriageIndex >= uint32(len(equivocationSet)) {
 				err := fmt.Errorf("marriage index out of bounds: %d > %d", post.MarriageIndex, len(equivocationSet)-1)
 				return nil, nil, err
 			}
+
 			id := equivocationSet[post.MarriageIndex]
 			effectiveNumUnits := post.NumUnits
 			if atx.Initial == nil {
 				var err error
 				effectiveNumUnits, err = h.validatePreviousAtx(id, &post, previousAtxs)
 				if err != nil {
-					return nil, nil, fmt.Errorf("validating previous atx for ID %s: %w", id, err)
+					return nil, nil, fmt.Errorf("validating previous atx: %w", err)
 				}
 			}
-			totalEffectiveNumUnits += effectiveNumUnits
+			nipostSizes[i].addUnits(effectiveNumUnits)
+		}
+	}
+
+	// validate poet membership proofs
+	for i, niposts := range atx.NiPosts {
+		// verify PoET memberships in a single go
+		indexedChallenges := make(map[uint64][]byte)
+
+		for _, post := range niposts.Posts {
+			nipostChallenge := wire.NIPostChallengeV2{
+				PublishEpoch:     atx.PublishEpoch,
+				PositioningATXID: atx.PositioningATX,
+			}
+			if atx.Initial != nil {
+				nipostChallenge.InitialPost = &atx.Initial.Post
+			} else {
+				nipostChallenge.PrevATXID = atx.PreviousATXs[post.PrevATXIndex]
+			}
+			if _, ok := indexedChallenges[post.MembershipLeafIndex]; !ok {
+				indexedChallenges[post.MembershipLeafIndex] = nipostChallenge.Hash().Bytes()
+			}
+		}
 
+		leafIndicies := make([]uint64, 0, len(indexedChallenges))
+		for i := range indexedChallenges {
+			leafIndicies = append(leafIndicies, i)
+		}
+		slices.Sort(leafIndicies)
+		poetChallenges := make([][]byte, 0, len(indexedChallenges))
+		for _, i := range leafIndicies {
+			poetChallenges = append(poetChallenges, indexedChallenges[i])
+		}
```

Comment:
That seems inefficient - `indexedChallenges` could be a `[][]byte` which would avoid the need for two copies and sorting `leafIndices` and `poetChallenges` 🙂

Reply:
I originally wanted to not use maps but slices and sort / compact them with the `slices` package. That doesn't seem to work as I thought, so instead you could use `maps.Keys` to save one manual range over the map:

```suggestion
		leafIndices := maps.Keys(indexedChallenges)
		slices.Sort(leafIndices)
		poetChallenges := make([][]byte, 0, len(indexedChallenges))
		for _, i := range leafIndices {
			poetChallenges = append(poetChallenges, indexedChallenges[i])
		}
```